src/bgw/job_stat.c

```diff
@@ -211,15 +212,37 @@ calculate_next_start_on_failure(TimestampTz finish_time, int consecutive_failure
 	BeginInternalSubTransaction("next start on failure");
 	PG_TRY();
 	{
-		/* ival = retry_period * (consecutive_failures)  */
-		Datum ival = DirectFunctionCall2(interval_mul,
-										 IntervalPGetDatum(&job->fd.retry_period),
-										 Float8GetDatum(multiplier));
+		int i;
+		Datum ival;
+		// max wait time to launch job is 1 minute
+		Interval interval_max = { .time = 60000000 };
+		Interval retry_ival = { .time = 2000000 };
+		/* NULL job means the failure was a launch failure */
+		if (job)
+			ival = IntervalPGetDatum(&job->fd.retry_period);
+		else
+		{
+			// retry every 2 seconds
+			ival = IntervalPGetDatum(&retry_ival);
+		}
+
+		/* ival = retry_period ^ (consecutive_failures - 1)  */
+		/* arbitrarily choose 2 for exponential growth */
+		for (i = 0; i < exponent - 1; i++)
+		{
+			ival = DirectFunctionCall2(interval_mul, ival, Float8GetDatum(2));
```

Comment:
Exponential backoff works when the scheduled time is picked randomly between `now()` and `now() + ival` (regardless of jitter). If we always pick `now() + ival` then we risk livelocking the conflicting jobs ad infinitum.

This is orthogonal to jitter, which helps with scheduling at non-integer multiples of time units, so as to not generate high load spikes in the system (e.g. due to everything happening at integer seconds instead of spreading them out).

Reply:
You are right, we are risking that. So for first case (out of background workers) I will change the backoff calculation to  randomly pick a backoff value between [0, 2^f - 1] where f is the number of consecutive failed launches of the job.